File: Front-End/src/services/voucher.service.ts

import axiosClient from '@/configurations/axios.config';
import type {
  CreateVoucherRequest,
  UpdateVoucherRequest,
  VoucherResponse,
  VoucherListResponse,
  VoucherAvailableResponse,
  Voucher
} from '@/types/voucher.type';

export const voucherService = {
  getVouchers: async (
    page: number = 1, 
    limit: number = 10, 
    name?: string, 
    type?: string, 
    active?: boolean, 
    startDate?: string, 
    endDate?: string
  ) => {
    const params = new URLSearchParams({
      page: page.toString(),
      limit: limit.toString(),
    });
    
    if (name) params.append('name', name);
    if (type) params.append('type', type);
    if (active !== undefined) params.append('active', active.toString());
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);
    
    const response = await axiosClient.get<VoucherListResponse>(`/vouchers?${params.toString()}`);
    return response.data;
  },

  getVoucherById: async (id: number) => {
    const response = await axiosClient.get<VoucherResponse>(`/vouchers/${id}`);
    return response.data;
  },

  createVoucher: async (request: CreateVoucherRequest) => {
    const response = await axiosClient.post<Voucher>('/vouchers', request);
    return response.data;
  },

  updateVoucher: async (id: number, request: UpdateVoucherRequest) => {
    const response = await axiosClient.put<Voucher>(`/vouchers/${id}`, request);
    return response.data;
  },

  changeStatusVoucher: async (id: number) => {
    await axiosClient.put(`/vouchers/change-status/${id}`);
  },

  sendVoucherToCustomers: async (id: number) => {
    await axiosClient.put(`/vouchers/${id}/send`);
  },

  getAvailableVouchers: async () => {
    const response = await axiosClient.get<VoucherAvailableResponse[]>('/vouchers/available');
    return response.data;
  }
};

File: Back-End/src/main/java/iuh/fit/ecommerce/services/VoucherService.java
package iuh.fit.ecommerce.services;


import iuh.fit.ecommerce.dtos.request.voucher.VoucherAddRequest;
import iuh.fit.ecommerce.dtos.request.voucher.VoucherUpdateRequest;
import iuh.fit.ecommerce.dtos.response.base.ResponseWithPagination;
import iuh.fit.ecommerce.dtos.response.voucher.VoucherAvailableResponse;
import iuh.fit.ecommerce.dtos.response.voucher.VoucherResponse;
import iuh.fit.ecommerce.entities.Voucher;
import jakarta.validation.Valid;

import java.time.LocalDate;
import java.util.List;

public interface VoucherService {
    VoucherResponse createVoucher( VoucherAddRequest request);

    VoucherResponse getVoucherById(Long id);

    ResponseWithPagination<List<VoucherResponse>> getAllVouchers(int page, int limit, String name, String type, Boolean active, LocalDate startDate, LocalDate endDate);

    VoucherResponse updateVoucher(Long id,  VoucherUpdateRequest request);

    void changeStatusVoucher(Long id);

    void sendVoucherToCustomers(Long id);

    Voucher getVoucherEntityById(Long id);

    List<VoucherAvailableResponse> getAvailableVouchersForCustomer();
}
